---
title: "striped_bass_msats_npop"
output: html_document
date: "2024-07-11"
---

```{r setup, include=FALSE}

if (!require(genepop)) install.packages('genepop')
if (!require(related)) install.packages("related", repos="http://R-Forge.R-project.org")
if (!require(ggplot2)) install.packages('ggplot2')
if (!require(dplyr)) install.packages('dplyr')
if (!require(BiocManager)) install.packages('BiocManager')
if (!require('adegenet')) install.packages('adegenet')
if (!require('pegas')) install.packages('pegas')
if (!require('janitor')) install.packages('janitor')
if (!require('tidytext')) install.packages("tidytext")
if (!require('hierfstat')) install.packages("hierfstat")
if (!require('pegas')) install.packages("pegas")
if (!require('devtools')) install.packages('devtools')
if (!require('graph4lg')) install.packages('graph4lg')
BiocManager::install("LEA")
if (!require(remotes)) install.packages('remotes')
remotes::install_github("rystanley/genepopedit")
if (!require(devtools)) install.packages('devtools')
devtools::install_github("kkeenan02/diveRsity")
install.packages(c("fields","RColorBrewer","mapplots"))
library('genepop')
library('related')
library('ggplot2')
library('dplyr')
library('adegenet')
library('genepopedit')
library('diveRsity')
library('LEA')
library('adegenet')
library('poppr')
library('dplyr')
library('hierfstat')
library('reshape2')
library('ggplot2')
library('RColorBrewer')
library('scales')
library('pegas')
library('graph4lg')
library('tidyverse')
library('StAMPP')

#Sets working directory
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

#sets a variable for your data file
locinfile <- "../data/fish/SBmicro_pops.txt"

```
```{r functions from GenePop2: null alleles, genotype frequencies, HW exact tests, linkage disequilibrium, IBD, private alleles}
##functions in GenePop2

#Null alleles
nulls(locinfile, outputFile = "../outputs/nulls.out")

#Allele and genotype frequencies per locus and per sample. 
basic_info(locinfile, outputFile = "../outputs/basic.out", verbose = interactive())

#Compute variants of the exact conditional test for Hardy-Weinberg genotypic proportions. 
test_HW(locinfile, which = "Proba", outputFile = "../outputs/SB_HW.out",
        settingsFile = "", enumeration = FALSE, dememorization = 10000,
        batches = 20, iterations = 5000, verbose = interactive())

#Exact test for genotypic linkage disequilibrium for each pair of loci in each population.
test_LD(locinfile, outputFile = "../outputs/SB_LD.out", settingsFile = "",
        dememorization = 10000, batches = 100, iterations = 5000,
        verbose = interactive())

#Evaluates Fst or related measures based on allele sizes, for all populations of for all pairs of populations.
Fst(locinfile, sizes = FALSE, pairs = TRUE, outputFile = "../outputs/SB_FST.out",
    dataType = "Diploid", verbose = interactive())

#note here that a special Genepop format is needed where every fish is a population and every sample name is coordinates

# Estimates isolation by distance by regression of genetic distance to geographical distance. 

###this needs to include geographic data
#File specification is for each individual to be a population and the name of each individual is its coordinates.
ibd('../data/fish/SBmicro_pops_dist.txt', '../outputs/SB_FST.out.MIG', dataType='diploid', geographicScale = 'Linear', statistic = 'SingleGeneDiv', mantelRankTest=TRUE)

#migration via private alleles
Nm_private(locinfile, outputFile = "../outputs/SBmicro_pops.txt.PRI ", dataType = "Diploid",
           verbose = interactive())
```

```{r functions from adegenet}
##generate genind object
sb_genind <- genepop_to_genind(locinfile, 14, pop_names = c("HR","DB","CB","NC","SC"))

#summarize genind object
sum_sb_genind <- (summary(sb_genind))
sum_sb_genind

#polymorphic loci
poly<-isPoly(sb_genind, by=c("locus"))
poly

#locus completeness
locmiss_sb <- propTyped(sb_genind, by = "loc")
barplot(locmiss_sb, ylim = c(0,1), ylab = "Complete genotypes (proportion)", xlab = "Locus", las = 2, cex.names = 0.7)

#private alleles
private_alleles(sb_genind) %>% apply(MARGIN = 1, FUN = sum)

#expected and observed heterozygosity
locus <- levels(sb_genind@loc.fac)
observed.hetero<- sum_sb_genind$Hobs
expected.hetero<- sum_sb_genind$Hexp 

my_table <- data.frame(locus = locus, observed.hetero = observed.hetero, expected.hetero = expected.hetero) |>
  mutate (deltaH = expected.hetero-observed.hetero)

table_long <- pivot_longer(my_table, c(expected.hetero,observed.hetero), names_to = "het")

het_plot <- ggplot(table_long, aes(x = locus)) +
  geom_bar(aes(y = value, fill=het), stat = "identity", position = position_dodge()) +
  theme(axis.text.x = element_text(size = 6),
panel.background = element_blank())

plot(het_plot)

#Fst values
sb_fst <- genet.dist(sb_genind, method = "WC84") %>% round(digits = 3)
sb_fst

#95% confidence intervals for global statistics
msat_hf <- genind2hierfstat(sb_genind)
boot.vc(msat_hf[1], msat_hf[-1])$ci
msat_hf 


#Discriminant analysis of Principal Components (DAPC)

find.clusters()

#Perform cross validation to find the optimal number of PCs to retain in DAPC
set.seed(123)
x <- tab(sb_genind, NA.method = "mean")
crossval = xvalDapc(x, sb_genind$pop, result = "groupMean", xval.plot = TRUE)

# Number of PCs with best stats (lower score = better)
crossval$`Root Mean Squared Error by Number of PCs of PCA`

crossval$`Number of PCs Achieving Highest Mean Success`

crossval$`Number of PCs Achieving Lowest MSE`

numPCs <- as.numeric(crossval$`Number of PCs Achieving Lowest MSE`)

#find number of genetic clusters
grp <- find.clusters(x, n.pca = numPCs, max.n = 12, n.clust = NULL)
plot(grp$Kstat, type="b", col="blue")

# Run a DAPC using site IDs as priors
dapc1 <- dapc(sb_genind, sb_genind$pop, n.pca = numPCs, n.da = 3)

# Analyse how much percent of genetic variance is explained by each axis
percent <- dapc1$eig/sum(dapc1$eig)*100
barplot(percent, ylab = "Genetic variance explained by eigenvectors (%)", ylim = c(0,60),
        names.arg = round(percent, 1))

# Create a data.frame containing individual coordinates
ind_coords <- as.data.frame(dapc1$ind.coord)

# Rename columns of dataframe
colnames(ind_coords) <- c("Axis1","Axis2","Axis3")

# Add a column containing individuals
ind_coords$Ind <- indNames(sb_genind)

# Add a column with the site IDs
ind_coords$Site <- sb_genind$pop

# Calculate centroid (average) position for each population
centroid <- aggregate(cbind(Axis1, Axis2, Axis3) ~ Site, data = ind_coords, FUN = mean)

# Add centroid coordinates to ind_coords dataframe
ind_coords <- left_join(ind_coords, centroid, by = "Site", suffix = c("",".cen"))

# Define colour palette
cols <- brewer.pal(nPop(sb_genind), "Set2")

# Custom x and y labels
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

# Scatter plot axis 1 vs. 2
ggplot(data = ind_coords, aes(x = Axis1, y = Axis2))+
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  # spider segments
  geom_segment(aes(xend = Axis1.cen, yend = Axis2.cen, colour = Site), show.legend = FALSE)+
  # points
  geom_point(aes(fill = Site), shape = 21, size = 3, show.legend = FALSE)+
  # centroids
  geom_label(data = centroid, aes(label = Site, fill = Site), size = 4, show.legend = FALSE)+
  # colouring
  scale_fill_manual(values = cols)+
  scale_colour_manual(values = cols)+
  # custom labels
  labs(x = xlab, y = ylab)+
  ggtitle("Striped bass DAPC")+
  # custom theme
  ggtheme

```
```{r diveRsity functions}
getwd()
save("../data/fish/SBmicro_pops.txt", file = "test2.rda")
test2<-load("test2.rda")
fastDivPart(Test_data, gp = 3, outfile = "testfolder")
```

```{r structure analysis}
struct2geno("../data/fish/SBmicro_pops.structure", 2, FORMAT = 2, extra.row = 0, extra.column = 2)

obj.snmf = snmf("../data/fish/SBmicro_pops.structure.geno", K = 1:8, ploidy = 2, entropy = T,
alpha = 100, project = "new")
plot(obj.snmf, col = "blue4", cex = 1.4, pch = 19)

obj.snmf = snmf("../data/fish/SBmicro_pops.structure.geno", K = 3, alpha = 100, project = "new")
qmatrix = Q(obj.snmf, K = 3)
barplot(t(qmatrix), col = c("red","blue","lightgreen"), border = NA, space = 0,
xlab = "Individuals", ylab = "Admixture coefficients")

```

```{r related}
#related analysis to determine pairwise relatedness

co2<-coancestry("SBmicro_pops.related2", wang=2, error.rates = 0.1)

#overall relatedness plot of coancestry output
ggplot(data = co2$relatedness, aes(x=group, y=wang)) + geom_point(shape=1)


#basic customizable way to output desired pairs
ggplot(data = subset(co2$relatedness, grepl("[2-3][890][2-3][890]", group)), aes(x=group, y=wang)) + geom_point(shape=1)

##only needed if dataset is making your R crash
#dat <- read.table("CA_related.5.txt", stringsAsFactors = FALSE, sep=" ", header = FALSE)
#dattrunc <- dat %>% select(1:401)
#co2trunc <- coancestry(dattrunc, wang=2)


sim <- familysim (co2$freqs , 100)
simoutput <- coancestry(sim, wang=2)
simrel <- cleanuprvals ( simoutput$relatedness , 100)
relvalues <- simrel [, 6]

label1 <- rep ("PO", 100)
label2 <- rep (" Full ", 100)
label3 <- rep (" Half ", 100)
label4 <- rep (" Unrelated ", 100)

labels <- c( label1 , label2 , label3 , label4 )

plot (as.factor(labels), relvalues , ylab ="Relatedness Value", xlab ="Relatedness")

# Use package adegenet: 
    #This package is devoted to the multivariate analysis of genetic markers data. These data can be codominant markers (e.g. microsatellites) or presence/absence data (e.g. AFLP), and have any level of ploidy. 'adegenet' defines three formal (S4) classes:
        #- genind: a class for data of individuals ("genind" stands for genotypes-individuals).
        #- genpop: a class for data of groups of individuals ("genpop" stands for genotypes-populations)
        #- genlight: a class for genome-wide SNP data
#For more information, visit the adegenet website using the function adegenetWeb.

#read in Genepop file (requires .gen extension)
SB_genind<-read.genepop("SBmicro_pops.gen", ncode = 3L, quiet = FALSE)

#clustering algorithm to find most probable number of clusters.  See documentation for details (?find.clusters)
SB.clust<-find.clusters(SB_genind, clust = NULL, n.pca = NULL, n.clust = NULL,
              method = c("kmeans", "ward"), stat = c("BIC","AIC", "WSS"),
              choose.n.clust = TRUE, criterion = c("diffNgroup", "min","goesup","smoothNgoesup", "goodfit"), 
              max.n.clust = round(nrow(SB_genind@tab)/10),
              n.iter = 1e5, n.start = 10, scale = FALSE, truenames = TRUE)

#principal components analysis of pre-defined groups (defined in Genepop file)
SB_dapc<-dapc(SB_genind, pop=NULL, n.pca=NULL, n.da=NULL, scale=FALSE,
              truenames=TRUE, var.contrib=TRUE, var.loadings=FALSE, pca.info=TRUE,
              pca.select=c("nbEig","percVar"), perc.pca=NULL)

#plotting for dapc object
scatter(SB_dapc)
assignplot(SB_dapc, only.grp=NULL, subset=NULL, new.pred=NULL, cex.lab=.75,pch=3)
compoplot(SB_dapc)

```


